File: frontend-app/src/app/services/direccion.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Direccion } from "../domain/Direccion";


@Injectable({
    providedIn: 'root'
})
export class DireccionService{

    private url:string = 'http://localhost:8080/api/direcciones';

    constructor(private http: HttpClient){

    }

    listadoDirecciones(): Observable<Direccion[]>{
        return this.http.get<Direccion[]>(this.url);
    }
    
        
    buscarDireccionPorId(id: number): Observable<Direccion>{
        return this.http.get<Direccion>(`${this.url}/${id}`);
    }

    guardarDireccion(direccion: Direccion){
        return this.http.post('http://localhost:8080/api/direcciones/guardar', direccion);
    }
    
    editarDireccion(direccion: Direccion) {
        return this.http.put(`${this.url}/actualizar`, direccion);
    }
    
    eliminarDireccion(idDireccion: number){
        return this.http.delete(`${this.url}/${idDireccion}`);
    }

    listadoDireccionPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }  

    buscarDireccionPorProvincia(provincia: string): Observable<Direccion[]>{
        const params = new HttpParams().set('provincia', provincia);
            
        return this.http.get<Direccion[]>(`${this.url}/buscar`, { params });
    }

    listadoDireccionesCompletas(): Observable<Direccion[]>{
        return this.http.get<Direccion[]>(`${this.url}/completas`);
    }

}

File: backend-app/src/main/java/com/backend_app/backend_app/service/DireccionService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.Direccion;
import com.backend_app.backend_app.dto.DireccionDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface DireccionService {

    List<DireccionDTO> listadoDirecciones();

    Optional<DireccionDTO> buscarDireccionPorId(long idDireccion);

    DireccionDTO convertirADTO(Direccion direccion);

    void insertarDireccion(DireccionDTO direccionDTO);

    void actualizarDireccion(DireccionDTO direccionDTO);

    void eliminarDireccion(DireccionDTO direccionDTO);

    Page<DireccionDTO> findAll(Pageable pageable);

    List<DireccionDTO> buscarDireccionPorProvincia(String provincia);

    List<DireccionDTO> listadoDireccionesCompletas();

}
